### ISSUE TITLE
[Connect 2.5] Workstreams - Assets tab

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3235]
On the work **Assets** tab we are showing files and links https://marvelapp.com/54cf844/screen/59214021

1. Where should they come from?
    At the moment I see two places where they can come from:
     - Comments on **Details** tab
     - Comments on **Requirements** tab
     - any other palce?

2. At the moment we are showing folders in "Assets Library" for topic/phase. But as this **Assets** tab is for the single Work, following such approach we wouldn't have any folders there. Or we can put file/link into the folders like **Datails**, **Requirements** and any other folders as per sources of files/links inside single work.

3. Should we have the ability to add files/links for works? Or only show them?

**Comment by vikasrohit** [comment]
1. Yes, I think those are the only two places (via files/links attache to the posts) from my understanding as well.

2. I think if we release with separate posts for Details and Requirements tab, we can have the name of the tabs (which are essentially going to be the name of the topics for those tabs as well) as folder names.

3. I think we can have the same Add New button in Assets tab as we have in the projects Assets Library.

While we are on this topic, I am not sure how would we handle creation of a folder in general for both of these places because right now we assume the name of the topic as a folder for all the links/files created in any post of the topic. However, we can definitely wait for new folder support until this release.

**Comment by maxceem** [comment]
Before we show all files and links extracted from Phases on the Assets Library page.

Now we don't have phases anymore and have works instead. And works have their own Assets tab.

Should we also show files and links extracted from work on the project Assets Library page as we did for phases before?

**Comment by vikasrohit** [comment]
Yes. The project's assets library should be the place where all project related assets can be located.

**Comment by maxceem** [comment]
We can do that. Though there is one thing: to show assets from work we have to load all the workstreams and all the works inside them and all the posts for works. Plus, if we already load all the works, for the consistency it's better to also load all timelines for works (though not required).

**Comment by maxceem** [comment]
On the second though, we also to load quite many stuff for phases. So I'll try to make it, will see how it works.

**Comment by vikasrohit** [comment]
Yep, that is something I am little worried about. The dashboard load is becoming very heavy and in fact it is already very heavy which in other sense is okay as dashboard is supposed to be convey all information about the project in one place. 😬 
Can we load only required data on dashboard before rendering the page and keep loading the other data after the page is loaded (making sure that the new data does not re-render the page)? I am not sure, how we can take care of this when the assets page is loaded directly, in which case we have to load the assets data before page rendering.

**Comment by maxceem** [comment]
> Can we load only required data on dashboard before rendering the page and keep loading the other data after the page is loaded (making sure that the new data does not re-render the page) 

Generally the Dashboard page is implemented this way now. Especially with new Workstreams functionality, data is loaded when required after the main page is already loaded.

So far we don't block showing Assets Library page while phases are loading. So files/links from phases just appear after phase are loaded in background.

For now we are always loading all the phases (without blocking UI in background). So we can do the same but only on Assets Library page for workstreams. And later we would think how we can improve it.

I don't think we should solve it now, as I guess soon we would have a bigger problem to show Assets Library page than loading all works. As the number of posts in project grows, I guess we would finally have to implement partial loading of posts (not all like now). And loading ALL posts for the project to extract files and links from them could become really heavy in some time, when we use posts more and more. So I guess we would have to implement some new endpoints in the Message Service to get all extracted links and files for a project. Probably by creating a separate ES index. Though we would also have take about permissions so users may only see files and links from the posts they can see... Ok, I think I have to stop here 😬What I wanted to say is, that probably we shouldn't optimize it now, as most likely soon, we should optimize it in another way as I've just described.

Probably I'm overthinking here, let me know what you think.

**Comment by maxceem** [comment]
**Update** I was wrong. We don't load phases and their topics on the **Assets Library** page. So we have an issue. If we directly access **Assets Library** page we don't see files and links from phases.

So we should fix this for both: phases and works.

**Comment by vikasrohit** [comment]
Yep, we need to fix the direct loading of the assets route/page.
> So far we don't block showing Assets Library page while phases are loading. So files/links from phases just appear after phase are loaded in background.

What I am worried with this approach is that assets page would show few files/links on page load and suddenly it would append few more as soon as the phases load. So, I guess, we need to show loading for the full page until we get all the data for the Assets page.

**Comment by vikasrohit** [comment]
On the idea of improving the logic (specially when we have many posts in a project) for files and links, I think files should not be a problem for us as we have a dedicated table `attachments` for that and it is easy to query a first class object separately. The only problem is the links which are not a first class object yet. Do you think we should make them first class object instead of nesting them inside the JSON field `bookmarks` in the project table?

**Comment by maxceem** [comment]
> On the idea of improving the logic (specially when we have many posts in a project) for files and links, I think files should not be a problem for us as we have a dedicated table attachments for that and it is easy to query a first class object separately. The only problem is the links which are not a first class object yet. Do you think we should make them first class object instead of nesting them inside the JSON field bookmarks in the project table?

Hmm, we don't have `bookmarks` for posts. We just parse posts content to extract links.
Also, regarding files inside posts. Even though we can retrieve attachments directly, we still have know their ids which are inside posts objects. 
So we still have to load all the posts to extract files and links from them.

**Comment by vikasrohit** [comment]
> Even though we can retrieve attachments directly, we still have to know their ids which are inside posts objects
Where in the post object we are storing the attachments ids?

**Comment by maxceem** [comment]
> Where in the post object we are storing the attachments ids?

Inside `attachments` array:

![image](https://user-images.githubusercontent.com/146016/63666951-94992680-c804-11e9-8416-d1b89e397570.png)

Also, inside that array we have attachment filename to show:

![image](https://user-images.githubusercontent.com/146016/63667142-3c165900-c805-11e9-8eb3-6f6d674f3180.png)

**Comment by vikasrohit** [comment]
Okay, Agreed that we have to pull the topics to show the attachments of the topics in the Assets library. I am bit worried about how we would handle it, as you mentioned also, when we have to implement paging for the topics. We can not always load all topics just to load the assets of those topics. I would discuss this with the team and come back to you for the future improvement, for now, we can keep fetching the posts before to load the assets.
